File: internal/handlers/items.go
package handlers

import (
	"database/sql"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	"materialvault/internal/db"
	"materialvault/internal/services"
	"materialvault/internal/utils"
)

// SyncItemTags ensures extracted tags exist and links them to item
func SyncItemTags(itemID string, explicitTagIDs []string, textToScan string) {
	extractedNames := utils.ExtractHashtags(textToScan)
	now := time.Now().UnixMilli()

	finalTagIDs := make(map[string]bool)
	for _, id := range explicitTagIDs {
		if strings.TrimSpace(id) != "" {
			finalTagIDs[id] = true
		}
	}

	for _, name := range extractedNames {
		var existingID string
		err := db.DB.QueryRow("SELECT id FROM tags WHERE name = ?", name).Scan(&existingID)
		if err == sql.ErrNoRows {
			newID := uuid.New().String()
			_, _ = db.DB.Exec("INSERT INTO tags (id, name, color, created_at) VALUES (?, ?, 'stone', ?)", newID, name, now)
			finalTagIDs[newID] = true
		} else if err == nil {
			finalTagIDs[existingID] = true
		}
	}

	for tagID := range finalTagIDs {
		_, _ = db.DB.Exec("INSERT OR IGNORE INTO item_tags (item_id, tag_id, created_at) VALUES (?, ?, ?)", itemID, tagID, now)
	}
}

// GetFullItem retrieves item with tags, assets, and logs
func GetFullItem(id string) (*db.Item, error) {
	var it db.Item
	var favoriteInt int
	var sourceURL, canonicalURL, sourceDomain sql.NullString
	var capturedAt sql.NullInt64

	row := db.DB.QueryRow(`
		SELECT id, type, title, description, source_url, canonical_url, source_domain, content_text, organization_status, processing_status, favorite, captured_at, created_at, updated_at
		FROM items
		WHERE id = ?
	`, id)

	err := row.Scan(
		&it.ID,
		&it.Type,
		&it.Title,
		&it.Description,
		&sourceURL,
		&canonicalURL,
		&sourceDomain,
		&it.ContentText,
		&it.OrganizationStatus,
		&it.ProcessingStatus,
		&favoriteInt,
		&capturedAt,
		&it.CreatedAt,
		&it.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	if sourceURL.Valid {
		it.SourceURL = &sourceURL.String
	}
	if canonicalURL.Valid {
		it.CanonicalURL = &canonicalURL.String
	}
	if sourceDomain.Valid {
		it.SourceDomain = &sourceDomain.String
	}
	if capturedAt.Valid {
		it.CapturedAt = &capturedAt.Int64
	}
	it.Favorite = favoriteInt == 1

	// Fetch Tags
	tagRows, err := db.DB.Query(`
		SELECT tg.id, tg.name, tg.color, tg.created_at
		FROM tags tg
		JOIN item_tags itg ON itg.tag_id = tg.id
		WHERE itg.item_id = ?
	`, id)
	it.Tags = []db.Tag{}
	if err == nil {
		defer tagRows.Close()
		for tagRows.Next() {
			var tg db.Tag
			if err := tagRows.Scan(&tg.ID, &tg.Name, &tg.Color, &tg.CreatedAt); err == nil {
				it.Tags = append(it.Tags, tg)
			}
		}
	}

	// Fetch Assets
	assetRows, err := db.DB.Query(`
		SELECT id, item_id, kind, mime_type, file_name, file_size, storage_path, sha256, created_at
		FROM assets
		WHERE item_id = ?
		ORDER BY created_at DESC
	`, id)
	it.Assets = []db.Asset{}
	if err == nil {
		defer assetRows.Close()
		for assetRows.Next() {
			var a db.Asset
			if err := assetRows.Scan(&a.ID, &a.ItemID, &a.Kind, &a.MimeType, &a.FileName, &a.FileSize, &a.StoragePath, &a.SHA256, &a.CreatedAt); err == nil {
				it.Assets = append(it.Assets, a)
			}
		}
	}

	// Fetch Logs
	logRows, err := db.DB.Query(`
		SELECT id, item_id, step, status, message, created_at
		FROM ingestion_logs
		WHERE item_id = ?
		ORDER BY created_at DESC
	`, id)
	it.Logs = []db.IngestionLog{}
	if err == nil {
		defer logRows.Close()
		for logRows.Next() {
			var l db.IngestionLog
			if err := logRows.Scan(&l.ID, &l.ItemID, &l.Step, &l.Status, &l.Message, &l.CreatedAt); err == nil {
				it.Logs = append(it.Logs, l)
			}
		}
	}

	return &it, nil
}

type CaptureURLRequest struct {
	URL         string   `json:"url"`
	Title       string   `json:"title"`
	Description string   `json:"description"`
	TagIDs      []string `json:"tagIds"`
}

type CreateNoteRequest struct {
	Title   string   `json:"title"`
	Content string   `json:"content"`
	TagIDs  []string `json:"tagIds"`
}

type BatchRequest struct {
	ItemIDs []string `json:"itemIds"`
	Action  string   `json:"action"` // set_status, favorite, unfavorite, delete, add_tag, remove_tag
	Status  string   `json:"status"`
	TagID   string   `json:"tagId"`
}

func HandleCaptureURL(assetsDir string) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		req, err := utils.ParseJSON[CaptureURLRequest](r)
		if err != nil || strings.TrimSpace(req.URL) == "" {
			utils.JSONError(w, http.StatusBadRequest, "无效请求参数")
			return
		}

		normalized, domain, err := utils.NormalizeURL(req.URL)
		if err != nil {
			utils.JSONError(w, http.StatusBadRequest, "URL 格式解析错误")
			return
		}

		// Check duplicate
		var existingID string
		err = db.DB.QueryRow("SELECT id FROM items WHERE canonical_url = ?", normalized).Scan(&existingID)
		if err == nil {
			SyncItemTags(existingID, req.TagIDs, req.Title+" "+req.Description)
			full, _ := GetFullItem(existingID)
			utils.JSON(w, http.StatusOK, map[string]any{"item": full, "isDuplicate": true})
			return
		}

		id := uuid.New().String()
		now := time.Now().UnixMilli()
		title := strings.TrimSpace(req.Title)
		if title == "" {
			title = normalized
		}
		description := strings.TrimSpace(req.Description)

		_, err = db.DB.Exec(`
			INSERT INTO items (id, type, title, description, source_url, canonical_url, source_domain, content_text, organization_status, processing_status, favorite, captured_at, created_at, updated_at)
			VALUES (?, 'url', ?, ?, ?, ?, ?, '', 'inbox', 'pending', 0, ?, ?, ?)
		`, id, title, description, req.URL, normalized, domain, now, now, now)

		if err != nil {
			utils.JSONError(w, http.StatusInternalServerError, "保存素材记录失败: "+err.Error())
			return
		}

		SyncItemTags(id, req.TagIDs, title+" "+description)

		// Async capture in background goroutine
		go func() {
			services.ProcessURLItem(assetsDir, id, normalized)
		}()

		full, _ := GetFullItem(id)
		utils.JSON(w, http.StatusAccepted, map[string]any{"item": full, "isDuplicate": false})
	}
}

func HandleCreateNote(w http.ResponseWriter, r *http.Request) {
	req, err := utils.ParseJSON[CreateNoteRequest](r)
	if err != nil || strings.TrimSpace(req.Content) == "" {
		utils.JSONError(w, http.StatusBadRequest, "请填写备忘内容")
		return
	}

	content := strings.TrimSpace(req.Content)
	lines := strings.Split(content, "\n")
	derivedTitle := strings.TrimSpace(req.Title)
	if derivedTitle == "" && len(lines) > 0 {
		if len(lines[0]) > 60 {
			derivedTitle = lines[0][:60]
		} else {
			derivedTitle = lines[0]
		}
	}
	if derivedTitle == "" {
		derivedTitle = "未命名备忘"
	}

	derivedDesc := ""
	if len(lines) > 1 {
		derivedDesc = strings.TrimSpace(strings.Join(lines[1:], "\n"))
	}

	id := uuid.New().String()
	now := time.Now().UnixMilli()

	_, err = db.DB.Exec(`
		INSERT INTO items (id, type, title, description, source_url, canonical_url, source_domain, content_text, organization_status, processing_status, favorite, captured_at, created_at, updated_at)
		VALUES (?, 'note', ?, ?, NULL, NULL, NULL, ?, 'inbox', 'ready', 0, ?, ?, ?)
	`, id, derivedTitle, derivedDesc, content, now, now, now)

	if err != nil {
		utils.JSONError(w, http.StatusInternalServerError, "保存备忘失败: "+err.Error())
		return
	}

	SyncItemTags(id, req.TagIDs, derivedTitle+" "+content)

	full, _ := GetFullItem(id)
	utils.JSON(w, http.StatusCreated, map[string]any{"item": full})
}

func HandleBatchItems(w http.ResponseWriter, r *http.Request) {
	req, err := utils.ParseJSON[BatchRequest](r)
	if err != nil || len(req.ItemIDs) == 0 {
		utils.JSONError(w, http.StatusBadRequest, "无效批量请求参数")
		return
	}

	placeholders := strings.Repeat("?,", len(req.ItemIDs))
	placeholders = placeholders[:len(placeholders)-1]
	now := time.Now().UnixMilli()

	var args []any
	for _, id := range req.ItemIDs {
		args = append(args, id)
	}

	switch req.Action {
	case "set_status":
		if req.Status != "" {
			updateArgs := append([]any{req.Status, now}, args...)
			_, _ = db.DB.Exec(fmt.Sprintf("UPDATE items SET organization_status = ?, updated_at = ? WHERE id IN (%s)", placeholders), updateArgs...)
		}
	case "favorite":
		updateArgs := append([]any{now}, args...)
		_, _ = db.DB.Exec(fmt.Sprintf("UPDATE items SET favorite = 1, updated_at = ? WHERE id IN (%s)", placeholders), updateArgs...)
	case "unfavorite":
		updateArgs := append([]any{now}, args...)
		_, _ = db.DB.Exec(fmt.Sprintf("UPDATE items SET favorite = 0, updated_at = ? WHERE id IN (%s)", placeholders), updateArgs...)
	case "delete":
		_, _ = db.DB.Exec(fmt.Sprintf("DELETE FROM items WHERE id IN (%s)", placeholders), args...)
	case "add_tag":
		if req.TagID != "" {
			for _, itemID := range req.ItemIDs {
				_, _ = db.DB.Exec("INSERT OR IGNORE INTO item_tags (item_id, tag_id, created_at) VALUES (?, ?, ?)", itemID, req.TagID, now)
			}
		}
	case "remove_tag":
		if req.TagID != "" {
			delArgs := append(args, req.TagID)
			_, _ = db.DB.Exec(fmt.Sprintf("DELETE FROM item_tags WHERE item_id IN (%s) AND tag_id = ?", placeholders), delArgs...)
		}
	}

	utils.JSON(w, http.StatusOK, map[string]any{"success": true, "count": len(req.ItemIDs)})
}

func HandleRetryItem(assetsDir string) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := r.PathValue("id")
		if id == "" {
			utils.JSONError(w, http.StatusBadRequest, "缺少 item id")
			return
		}

		var itType, sourceURL, canonicalURL sql.NullString
		err := db.DB.QueryRow("SELECT type, source_url, canonical_url FROM items WHERE id = ?", id).Scan(&itType, &sourceURL, &canonicalURL)
		if err != nil {
			utils.JSONError(w, http.StatusNotFound, "素材不存在")
			return
		}
		if itType.String != "url" || (!sourceURL.Valid && !canonicalURL.Valid) {
			utils.JSONError(w, http.StatusBadRequest, "仅支持 URL 类型的素材重新抓取")
			return
		}

		targetURL := canonicalURL.String
		if targetURL == "" {
			targetURL = sourceURL.String
		}

		go func() {
			services.ProcessURLItem(assetsDir, id, targetURL)
		}()

		utils.JSON(w, http.StatusOK, map[string]string{"message": "重新归档任务已提交"})
	}
}

func HandleListItems(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	status := q.Get("status")
	itemType := q.Get("type")
	tagID := q.Get("tagId")
	domain := q.Get("domain")
	favStr := q.Get("favorite")

	var favorite *bool
	if favStr == "true" || favStr == "1" {
		b := true
		favorite = &b
	} else if favStr == "false" || favStr == "0" {
		b := false
		favorite = &b
	}

	limit := 50
	if l, err := strconv.Atoi(q.Get("limit")); err == nil && l > 0 {
		limit = l
	}

	offset := 0
	if o, err := strconv.Atoi(q.Get("offset")); err == nil && o >= 0 {
		offset = o
	}

	var startDate, endDate *int64
	if s, err := strconv.ParseInt(q.Get("startDate"), 10, 64); err == nil {
		startDate = &s
	}
	if e, err := strconv.ParseInt(q.Get("endDate"), 10, 64); err == nil {
		endDate = &e
	}

	items, err := services.SearchItems(services.SearchFilters{
		Q:                  q.Get("q"),
		Type:               itemType,
		OrganizationStatus: status,
		TagID:              tagID,
		Domain:             domain,
		Favorite:           favorite,
		StartDate:          startDate,
		EndDate:            endDate,
		Limit:              limit,
		Offset:             offset,
	})

	if err != nil {
		utils.JSONError(w, http.StatusInternalServerError, "获取素材列表失败: "+err.Error())
		return
	}

	if items == nil {
		items = []db.Item{}
	}

	utils.JSON(w, http.StatusOK, map[string]any{"items": items})
}

func HandleGetItem(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	item, err := GetFullItem(id)
	if err != nil {
		utils.JSONError(w, http.StatusNotFound, "素材不存在")
		return
	}
	utils.JSON(w, http.StatusOK, map[string]any{"item": item})
}

type UpdateItemRequest struct {
	Title              *string `json:"title"`
	Description        *string `json:"description"`
	ContentText        *string `json:"contentText"`
	OrganizationStatus *string `json:"organizationStatus"`
	Favorite           *bool   `json:"favorite"`
}

func HandleUpdateItem(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	req, err := utils.ParseJSON[UpdateItemRequest](r)
	if err != nil {
		utils.JSONError(w, http.StatusBadRequest, "无效更新参数")
		return
	}

	var setClauses []string
	var args []any
	now := time.Now().UnixMilli()

	if req.Title != nil {
		setClauses = append(setClauses, "title = ?")
		args = append(args, *req.Title)
	}
	if req.Description != nil {
		setClauses = append(setClauses, "description = ?")
		args = append(args, *req.Description)
	}
	if req.ContentText != nil {
		setClauses = append(setClauses, "content_text = ?")
		args = append(args, *req.ContentText)
	}
	if req.OrganizationStatus != nil {
		setClauses = append(setClauses, "organization_status = ?")
		args = append(args, *req.OrganizationStatus)
	}
	if req.Favorite != nil {
		setClauses = append(setClauses, "favorite = ?")
		if *req.Favorite {
			args = append(args, 1)
		} else {
			args = append(args, 0)
		}
	}

	if len(setClauses) > 0 {
		setClauses = append(setClauses, "updated_at = ?")
		args = append(args, now)
		args = append(args, id)

		query := fmt.Sprintf("UPDATE items SET %s WHERE id = ?", strings.Join(setClauses, ", "))
		_, err = db.DB.Exec(query, args...)
		if err != nil {
			utils.JSONError(w, http.StatusInternalServerError, "更新素材失败: "+err.Error())
			return
		}
	}

	// Auto extract hashtags from updated content
	var fullText string
	if req.Title != nil {
		fullText += *req.Title + " "
	}
	if req.Description != nil {
		fullText += *req.Description + " "
	}
	if req.ContentText != nil {
		fullText += *req.ContentText
	}
	if strings.Contains(fullText, "#") {
		SyncItemTags(id, nil, fullText)
	}

	full, _ := GetFullItem(id)
	utils.JSON(w, http.StatusOK, map[string]any{"item": full})
}

func HandleDeleteItem(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	_, err := db.DB.Exec("DELETE FROM items WHERE id = ?", id)
	if err != nil {
		utils.JSONError(w, http.StatusInternalServerError, "删除失败: "+err.Error())
		return
	}
	utils.JSON(w, http.StatusOK, map[string]any{"success": true, "id": id})
}

type LinkTagRequest struct {
	TagID   string `json:"tagId"`
	TagName string `json:"tagName"`
}

func HandleLinkItemTag(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	req, err := utils.ParseJSON[LinkTagRequest](r)
	if err != nil {
		utils.JSONError(w, http.StatusBadRequest, "无效参数")
		return
	}

	targetTagID := req.TagID
	now := time.Now().UnixMilli()

	if targetTagID == "" && req.TagName != "" {
		cleaned := strings.TrimPrefix(strings.TrimSpace(req.TagName), "#")
		var existingID string
		err := db.DB.QueryRow("SELECT id FROM tags WHERE name = ?", cleaned).Scan(&existingID)
		if err == sql.ErrNoRows {
			newID := uuid.New().String()
			_, _ = db.DB.Exec("INSERT INTO tags (id, name, color, created_at) VALUES (?, ?, 'stone', ?)", newID, cleaned, now)
			targetTagID = newID
		} else if err == nil {
			targetTagID = existingID
		}
	}

	if targetTagID == "" {
		utils.JSONError(w, http.StatusBadRequest, "缺少 tagId 或 tagName")
		return
	}

	_, _ = db.DB.Exec("INSERT OR IGNORE INTO item_tags (item_id, tag_id, created_at) VALUES (?, ?, ?)", id, targetTagID, now)

	full, _ := GetFullItem(id)
	utils.JSON(w, http.StatusOK, map[string]any{"item": full})
}

func HandleUnlinkItemTag(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	tagID := r.PathValue("tagId")

	_, _ = db.DB.Exec("DELETE FROM item_tags WHERE item_id = ? AND tag_id = ?", id, tagID)

	full, _ := GetFullItem(id)
	utils.JSON(w, http.StatusOK, map[string]any{"item": full})
}

File: server/routes/items.ts
import { z } from 'zod';
import crypto from 'crypto';
import { db, sqlite } from '../db/db';
import { items, itemTags, assets, ingestionLogs, tags } from '../db/schema';
import { eq, desc, inArray } from 'drizzle-orm';
import { normalizeUrl, processUrlItem } from '../services/capture';
import { json, parseJsonBody } from '../utils';

// Helper to extract hashtags from any text: #标签名 #tech #ai_123
export function extractHashtags(text: string): string[] {
  if (!text) return [];
  const matches = text.match(/(?:^|\s)#([\p{L}\p{N}_-]+)/gu);
  if (!matches) return [];

  const tagSet = new Set<string>();
  for (const m of matches) {
    const cleaned = m.trim().replace(/^#/, '').trim();
    if (cleaned.length > 0) {
      tagSet.add(cleaned);
    }
  }
  return Array.from(tagSet);
}

// Helper to ensure tag exists and link to item
export async function syncItemTags(itemId: string, explicitTagIds: string[] = [], textToScan = '') {
  const extractedTagNames = extractHashtags(textToScan);
  const now = Date.now();
  const finalTagIds = new Set<string>(explicitTagIds.filter(Boolean));

  for (const name of extractedTagNames) {
    let existing = await db.query.tags.findFirst({
      where: eq(tags.name, name),
    });

    if (!existing) {
      const newTagId = crypto.randomUUID();
      await db.insert(tags).values({
        id: newTagId,
        name,
        color: 'stone',
        createdAt: now,
      });
      finalTagIds.add(newTagId);
    } else {
      finalTagIds.add(existing.id);
    }
  }

  // Insert links
  for (const tagId of finalTagIds) {
    try {
      await db.insert(itemTags).values({ itemId, tagId, createdAt: now });
    } catch (_) {}
  }
}

// Helper to fetch item with associated tags, assets, and logs
export async function getFullItem(id: string) {
  const item = await db.query.items.findFirst({
    where: eq(items.id, id),
  });
  if (!item) return null;

  // Fetch tags
  const tagList = sqlite
    .prepare(
      `SELECT tg.* FROM tags tg JOIN item_tags itg ON itg.tag_id = tg.id WHERE itg.item_id = ?`
    )
    .all(id);

  // Fetch assets
  const assetList = await db.query.assets.findMany({
    where: eq(assets.itemId, id),
    orderBy: [desc(assets.createdAt)],
  });

  // Fetch logs
  const logList = await db.query.ingestionLogs.findMany({
    where: eq(ingestionLogs.itemId, id),
    orderBy: [desc(ingestionLogs.createdAt)],
  });

  return {
    ...item,
    favorite: Boolean(item.favorite),
    tags: tagList,
    assets: assetList,
    logs: logList,
  };
}

export async function handleItems(req: Request, url: URL, subPath: string): Promise<Response> {
  const method = req.method;

  // 1. Capture URL: POST /api/items/url
  if (method === 'POST' && subPath === 'url') {
    const body = await parseJsonBody(req);
    const schema = z.object({
      url: z.string().min(1),
      title: z.string().optional(),
      description: z.string().optional(),
      tagIds: z.array(z.string()).optional(),
    });

    const parsed = schema.safeParse(body);
    if (!parsed.success) {
      return json({ error: '无效请求参数', details: parsed.error.format() }, 400);
    }

    const { url: rawUrl, title, description, tagIds } = parsed.data;
    const { url: normalized, domain } = normalizeUrl(rawUrl);

    // Check duplicate URL
    const existing = await db.query.items.findFirst({
      where: eq(items.canonicalUrl, normalized),
    });

    if (existing) {
      await syncItemTags(existing.id, tagIds || [], `${title || ''} ${description || ''}`);
      const full = await getFullItem(existing.id);
      return json({ item: full, isDuplicate: true }, 200);
    }

    const id = crypto.randomUUID();
    const now = Date.now();

    const newItem = {
      id,
      type: 'url' as const,
      title: title?.trim() || normalized,
      description: description?.trim() || '',
      sourceUrl: rawUrl,
      canonicalUrl: normalized,
      sourceDomain: domain,
      contentText: '',
      organizationStatus: 'inbox' as const,
      processingStatus: 'pending' as const,
      favorite: false,
      capturedAt: now,
      createdAt: now,
      updatedAt: now,
    };

    await db.insert(items).values(newItem);

    // Link & extract tags from text
    await syncItemTags(id, tagIds || [], `${title || ''} ${description || ''}`);

    setTimeout(() => {
      processUrlItem(id, normalized).catch((err) => console.error('Capture pipeline error:', err));
    }, 10);

    const full = await getFullItem(id);
    return json({ item: full, isDuplicate: false }, 202);
  }

  // 2. Create Note: POST /api/items/note
  if (method === 'POST' && subPath === 'note') {
    const body = await parseJsonBody(req);
    const schema = z.object({
      title: z.string().optional(),
      content: z.string().min(1),
      tagIds: z.array(z.string()).optional(),
    });

    const parsed = schema.safeParse(body);
    if (!parsed.success) {
      return json({ error: '请填写备忘内容', details: parsed.error.format() }, 400);
    }

    const { title, content, tagIds } = parsed.data;
    const id = crypto.randomUUID();
    const now = Date.now();

    const lines = content.trim().split('\n');
    const derivedTitle = title?.trim() || lines[0].slice(0, 60) || '未命名备忘';
    const derivedDesc = lines.length > 1 ? lines.slice(1).join('\n').trim() : '';

    const newItem = {
      id,
      type: 'note' as const,
      title: derivedTitle,
      description: derivedDesc,
      sourceUrl: null,
      canonicalUrl: null,
      sourceDomain: null,
      contentText: content.trim(),
      organizationStatus: 'inbox' as const,
      processingStatus: 'ready' as const,
      favorite: false,
      capturedAt: now,
      createdAt: now,
      updatedAt: now,
    };

    await db.insert(items).values(newItem);

    // Auto extract #tags from entire content & title
    await syncItemTags(id, tagIds || [], `${derivedTitle} ${content}`);

    const full = await getFullItem(id);
    return json({ item: full }, 201);
  }

  // 3. Batch operations: POST /api/items/batch
  if (method === 'POST' && subPath === 'batch') {
    const body = await parseJsonBody(req);
    const schema = z.object({
      itemIds: z.array(z.string()).min(1),
      action: z.enum(['set_status', 'add_tag', 'remove_tag', 'favorite', 'unfavorite', 'delete']),
      status: z.enum(['inbox', 'organized', 'archived']).optional(),
      tagId: z.string().optional(),
    });

    const parsed = schema.safeParse(body);
    if (!parsed.success) {
      return json({ error: '无效批量请求参数' }, 400);
    }

    const { itemIds, action, status, tagId } = parsed.data;

    switch (action) {
      case 'set_status':
        if (status) {
          await db.update(items).set({ organizationStatus: status, updatedAt: Date.now() }).where(inArray(items.id, itemIds));
        }
        break;
      case 'favorite':
        await db.update(items).set({ favorite: true, updatedAt: Date.now() }).where(inArray(items.id, itemIds));
        break;
      case 'unfavorite':
        await db.update(items).set({ favorite: false, updatedAt: Date.now() }).where(inArray(items.id, itemIds));
        break;
      case 'delete':
        await db.delete(items).where(inArray(items.id, itemIds));
        break;
      case 'add_tag':
        if (tagId) {
          for (const itemId of itemIds) {
            try {
              await db.insert(itemTags).values({ itemId, tagId, createdAt: Date.now() });
            } catch (_) {}
          }
        }
        break;
      case 'remove_tag':
        if (tagId) {
          await db.delete(itemTags).where(inArray(itemTags.itemId, itemIds));
        }
        break;
    }

    return json({ success: true, count: itemIds.length });
  }

  // 4. Retry Ingestion: POST /api/items/:id/retry
  if (method === 'POST' && subPath.endsWith('/retry')) {
    const id = subPath.replace(/\/retry$/, '');
    const item = await db.query.items.findFirst({ where: eq(items.id, id) });
    if (!item) return json({ error: '素材不存在' }, 404);
    if (item.type !== 'url' || !item.sourceUrl) {
      return json({ error: '仅支持 URL 类型的素材重新抓取' }, 400);
    }

    setTimeout(() => {
      processUrlItem(id, item.canonicalUrl || item.sourceUrl!).catch(console.error);
    }, 10);

    return json({ message: '重新归档任务已提交' });
  }

  // 5. Link / Unlink Tag: POST/DELETE /api/items/:id/tags...
  if (subPath.includes('/tags')) {
    const parts = subPath.split('/tags');
    const itemId = parts[0];
    const rest = parts[1] ? parts[1].replace(/^\//, '') : '';

    if (method === 'POST') {
      const body = await parseJsonBody(req);
      let targetTagId = body?.tagId;
      const tagName = body?.tagName;

      if (!targetTagId && tagName) {
        const cleanedName = tagName.trim().replace(/^#/, '');
        const existing = await db.query.tags.findFirst({ where: eq(tags.name, cleanedName) });
        if (existing) {
          targetTagId = existing.id;
        } else {
          targetTagId = crypto.randomUUID();
          await db.insert(tags).values({
            id: targetTagId,
            name: cleanedName,
            color: 'stone',
            createdAt: Date.now(),
          });
        }
      }

      if (!targetTagId) return json({ error: '缺少 tagId 或 tagName' }, 400);

      try {
        await db.insert(itemTags).values({ itemId, tagId: targetTagId, createdAt: Date.now() });
      } catch (_) {}

      const full = await getFullItem(itemId);
      return json({ item: full });
    }

    if (method === 'DELETE' && rest) {
      const tagId = rest;
      const { and } = await import('drizzle-orm');
      await db.delete(itemTags).where(and(eq(itemTags.itemId, itemId), eq(itemTags.tagId, tagId)));
      const full = await getFullItem(itemId);
      return json({ item: full });
    }
  }

  // 6. Get single Item: GET /api/items/:id
  if (method === 'GET' && subPath) {
    const item = await getFullItem(subPath);
    if (!item) return json({ error: '素材不存在' }, 404);
    return json({ item });
  }

  // 7. Update Item: PATCH /api/items/:id
  if (method === 'PATCH' && subPath) {
    const body = await parseJsonBody(req);
    const schema = z.object({
      title: z.string().optional(),
      description: z.string().optional(),
      contentText: z.string().optional(),
      organizationStatus: z.enum(['inbox', 'organized', 'archived']).optional(),
      favorite: z.boolean().optional(),
    });

    const parsed = schema.safeParse(body);
    if (!parsed.success) {
      return json({ error: '无效更新参数' }, 400);
    }

    const updates: any = {
      ...parsed.data,
      updatedAt: Date.now(),
    };

    await db.update(items).set(updates).where(eq(items.id, subPath));

    // Auto extract #tags from updated title/description/content
    const textToScan = `${parsed.data.title || ''} ${parsed.data.description || ''} ${parsed.data.contentText || ''}`;
    if (textToScan.includes('#')) {
      await syncItemTags(subPath, [], textToScan);
    }

    const full = await getFullItem(subPath);
    return json({ item: full });
  }

  // 8. Delete Item: DELETE /api/items/:id
  if (method === 'DELETE' && subPath) {
    await db.delete(items).where(eq(items.id, subPath));
    return json({ success: true, id: subPath });
  }

  // 9. List items: GET /api/items
  if (method === 'GET' && !subPath) {
    const status = url.searchParams.get('status') as any;
    const type = url.searchParams.get('type');
    const favorite = url.searchParams.get('favorite');
    const tagId = url.searchParams.get('tagId');
    const limit = parseInt(url.searchParams.get('limit') || '50', 10);
    const offset = parseInt(url.searchParams.get('offset') || '0', 10);

    let querySql = `SELECT DISTINCT i.* FROM items i`;
    const joins: string[] = [];
    const where: string[] = [];
    const params: any[] = [];

    if (tagId) {
      joins.push('JOIN item_tags itg ON itg.item_id = i.id');
      where.push('itg.tag_id = ?');
      params.push(tagId);
    }

    if (status) {
      where.push('i.organization_status = ?');
      params.push(status);
    }

    if (type) {
      where.push('i.type = ?');
      params.push(type);
    }

    if (favorite === 'true' || favorite === '1') {
      where.push('i.favorite = 1');
    }

    querySql += ' ' + joins.join(' ');
    if (where.length > 0) {
      querySql += ' WHERE ' + where.join(' AND ');
    }
    querySql += ' ORDER BY i.created_at DESC LIMIT ? OFFSET ?';
    params.push(limit, offset);

    const rawItems = sqlite.prepare(querySql).all(...params) as any[];
    const itemIds = rawItems.map((r) => r.id);
    let tagsByItem: Record<string, any[]> = {};
    let assetsByItem: Record<string, any[]> = {};

    if (itemIds.length > 0) {
      const placeholders = itemIds.map(() => '?').join(',');

      const tagLinks = sqlite
        .prepare(
          `SELECT itg.item_id, tg.* FROM tags tg JOIN item_tags itg ON itg.tag_id = tg.id WHERE itg.item_id IN (${placeholders})`
        )
        .all(...itemIds) as any[];
      tagLinks.forEach((tg) => {
        if (!tagsByItem[tg.item_id]) tagsByItem[tg.item_id] = [];
        tagsByItem[tg.item_id].push({ id: tg.id, name: tg.name, color: tg.color });
      });

      const assetRows = sqlite
        .prepare(
          `SELECT id, item_id, kind, mime_type, file_name, file_size, storage_path, created_at FROM assets WHERE item_id IN (${placeholders})`
        )
        .all(...itemIds) as any[];
      assetRows.forEach((a) => {
        if (!assetsByItem[a.item_id]) assetsByItem[a.item_id] = [];
        assetsByItem[a.item_id].push(a);
      });
    }

    const result = rawItems.map((r) => ({
      id: r.id,
      type: r.type,
      title: r.title,
      description: r.description,
      sourceUrl: r.source_url,
      canonicalUrl: r.canonical_url,
      sourceDomain: r.source_domain,
      contentText: r.content_text,
      organizationStatus: r.organization_status,
      processingStatus: r.processing_status,
      favorite: Boolean(r.favorite),
      capturedAt: r.captured_at,
      createdAt: r.created_at,
      updatedAt: r.updated_at,
      tags: tagsByItem[r.id] || [],
      assets: assetsByItem[r.id] || [],
    }));

    return json({ items: result });
  }

  return json({ error: 'Not Found' }, 404);
}
